### PR TITLE
Extract vindex lookup queries into their own primitive

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description
This PR changes how we plan vindex lookup queries. Before this change, the vindex lookup query was done by the vindex at runtime. This means that at plan time, we still haven't planned the lookup query and can't show it in plan descriptions.

With this change, vindex lookups are performed by a new engine primitive, the `VindexLookup`. It looks like a join - it has two inputs, and fetches data from the left-hand side query to figure out where to send the right-hand side query.

This makes the lookup query explicit in the plan, helping see the full scope of a query. This can be important to optimize queries as well as possible.

As an example, here is the plan for `select id from user where name = 'bb'`:
```json
{
  "OperatorType": "VindexLookup",
  "Variant": "Equal",
  "Keyspace": {
    "Name": "user",
    "Sharded": true
  },
  "Values": [
    "VARCHAR(\"bb\")"
  ],
  "Vindex": "name_user_map",
  "Inputs": [
    {
      "OperatorType": "Route",
      "Variant": "EqualUnique",
      "Keyspace": {
        "Name": "user",
        "Sharded": true
      },
      "FieldQuery": "select name, id from user_name_lookup where 1 != 1",
      "Query": "select name, id from user_name_lookup where name in ::col1",
      "Table": "`user`"
    },
    {
      "OperatorType": "Route",
      "Variant": "ByDestination",
      "Keyspace": {
        "Name": "user",
        "Sharded": true
      },
      "FieldQuery": "select id from `user` where 1 != 1",
      "Query": "select id from `user` where `name` = 'bb'",
      "Table": "`user`"
    }
  ]
}
```

In the plan above, you can see the first route, which is the query used to fetch data from the lookup vindex,

 ## Left todo:
- [x] Streaming implementation
- [x] Work on plan_test to make more sense. Lookup queries should not scatter
- [x] Add e2e with dirty transaction state - inserts, deletes and updates
- [x] Find more corner cases and failure cases
- [x] Implement PlanableVindex for more vindexes

## Checklist
-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/vtgate/lookup_queries/main_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2021 The Vitess Authors.
```
Comment (systay) [inline go/test/endtoend/vtgate/lookup_queries/main_test.go:2]: 2022

File: go/test/endtoend/vtgate/lookup_queries/main_test.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	"vitess.io/vitess/go/test/endtoend/cluster"
+)
+
+var (
+	clusterInstance *cluster.LocalProcessCluster
+	vtParams        mysql.ConnParams
+	mysqlParams     mysql.ConnParams
+	shardedKs       = "ks"
+
+	shardedKsShards = []string{"-19a0", "19a0-20", "20-20c0", "20c0-"}
```
Comment (systay) [inline go/test/endtoend/vtgate/lookup_queries/main_test.go:42]: Y WEIRD?

Comment (harshit-gangal) [reply go/test/endtoend/vtgate/lookup_queries/main_test.go:42]: to show the benefit of subsharding vindex.

File: go/test/endtoend/vtgate/lookup_queries/main_test.go (hunk 1 above)
Comment (harshit-gangal) [reply go/test/endtoend/vtgate/lookup_queries/main_test.go:2]: :+1:

File: go/vt/vtexplain/testdata/multi-output/selectsharded-output.txt [hunk 3]
```diff
@@ -59,17 +59,11 @@ select count(*) from user where id = 1 /* point aggregate */
...
-8 ks_sharded/40-80: select `name`, user_id from name_user_map where `name` in ('h') limit 10001 /* scatter aggregate */
-9 ks_sharded/-40: select `name`, user_id from name_user_map where `name` in ('i') limit 10001 /* scatter aggregate */
-10 ks_sharded/-40: select `name`, user_id from name_user_map where `name` in ('j') limit 10001 /* scatter aggregate */
-11 ks_sharded/-40: select count(*) from `user` where `name` in ('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j') limit 10001 /* scatter aggregate */
+1 ks_sharded/-40: select `name`, user_id from name_user_map where `name` in ('f', 'g', 'i', 'j') limit 10001 /* scatter aggregate */
+1 ks_sharded/40-80: select `name`, user_id from name_user_map where `name` in ('b', 'c', 'h') limit 10001 /* scatter aggregate */
+1 ks_sharded/80-c0: select `name`, user_id from name_user_map where `name` in ('e') limit 10001 /* scatter aggregate */
+1 ks_sharded/c0-: select `name`, user_id from name_user_map where `name` in ('a', 'd') limit 10001 /* scatter aggregate */
+2 ks_sharded/-40: select count(*) from `user` where `name` in ('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j') limit 10001 /* scatter aggregate */
```
Comment (harshit-gangal) [inline go/vt/vtexplain/testdata/multi-output/selectsharded-output.txt:66]: this should not be batches as they are varchar and hence not comparable unless the collation is known.

File: go/vt/vtgate/engine/vindex_lookup.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+	dest, err := vr.postLookupPreExecute(ids, results, bindVars)
+	if err != nil {
+		return nil, err
+	}
+
+	return vr.SendTo.executeAfterLookup(ctx, vcursor, bindVars, wantfields, ids, dest)
+}
+
+func (vr *VindexLookup) postLookupPreExecute(ids []sqltypes.Value, results []*sqltypes.Result, bindVars map[string]*querypb.BindVariable) ([]key.Destination, error) {
```
Comment (systay) [inline go/vt/vtgate/engine/vindex_lookup.go:101]: nit: this could be named differently - `mapVindexValues` or something would be more descriptive

File: go/vt/vtgate/vindexes/vindex.go [hunk 5]
```diff
@@ -28,130 +28,138 @@ import (
...
-// A VCursor is an interface that allows you to execute queries
-// in the current context and session of a VTGate request. Vindexes
-// can use this interface to execute lookup queries.
-type VCursor interface {
-	Execute(ctx context.Context, method string, query string, bindvars map[string]*querypb.BindVariable, rollbackOnError bool, co vtgatepb.CommitOrder) (*sqltypes.Result, error)
-	ExecuteKeyspaceID(ctx context.Context, keyspace string, ksid []byte, query string, bindVars map[string]*querypb.BindVariable, rollbackOnError, autocommit bool) (*sqltypes.Result, error)
-	InTransactionAndIsDML() bool
-	LookupRowLockShardSession() vtgatepb.CommitOrder
-}
+) // This file defines interfaces and registration for vindexes.
```
Comment (systay) [inline go/vt/vtgate/vindexes/vindex.go:31]: nit: this comment seems to have ended up on the wrong line

**Comment by systay** [comment]
LGTM. I'll let someone else ACK it though

**Review by frouioui — APPROVED** [review]

**Comment by frouioui** [comment]
Approving this PR as @systay has already ack this

File: go/vt/vtgate/engine/primitive.go [hunk 6]
```diff
@@ -172,6 +172,8 @@ type (
 
 		VtExplainLogging()
 		GetVTExplainLogs() []ExecuteEntry
+
+		SetCommitOrder(co vtgatepb.CommitOrder)
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/engine/primitive.go:176]: @harshit-gangal can you do this? Not sure that I can explain it well
